### ISSUE TITLE
Azure App Service deploy started failing with Unrecognized argument 'Files\IIS\Microsoft'. All arguments must begin with "-"

### ISSUE THREAD

**Issue by noellelc** [microsoft/azure-pipelines-tasks#17634]
## Note  
Issues in this repo are for tracking bugs, feature requests and questions for the tasks in this repo 

For a list:  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

If you have an issue or request for the Azure Pipelines service, use developer community instead:  

https://developercommunity.visualstudio.com/spaces/21/index.html )

## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: bug

**Enter Task Name**: Azure App Service deploy

list here (V# not needed):  4.216.2
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment
- Server - Azure Pipelines or TFS on-premises?
    
    - If using TFS on-premises, provide the version: 
    
    - If using Azure Pipelines, provide the account name, team project name, build definition name/build number: 
       account name: DevDiv, Project name: DevDiv, Release id: 103259

- Agent - Hosted or Private: 
    
    - If using Hosted agent, provide agent queue name: private

    - If using private agent, provide the OS of the machine running the agent and the agent version: 

## Issue Description

We are running this task in a release pipeline in Azure DevOps. Last successful run was yesterday on version 4.214.0. It looks like this task just recently started failing, with version 4.216.2. Error message: Unrecognized argument 'Files\IIS\Microsoft'. All arguments must begin with "-".

The argument in question is the command itself: "C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe"; it looks like it's not honoring the quotations and is parsing this string as individual arguments, hence the callout of "Files\IIS\Microsoft"

![image](https://user-images.githubusercontent.com/63023800/213494434-1ae97ac5-b77e-4399-a040-e85e922b79ac.png)

### Task logs

[Enable debug logging and please provide the zip file containing all the logs for a speedy resolution]

## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

### Error logs

[Insert error from the logs here for a quick overview]

**Comment by tassioFront** [comment]
I just have gotten it too

**Comment by rafjsouza** [comment]
Same here

**Comment by guiifva** [comment]
Same here

**Comment by warlima** [comment]
Same here

**Comment by AlejandroReinoza** [comment]
Same here

**Comment by jmendezdd** [comment]
Same here

**Comment by blogcraft** [comment]
Same...

**Comment by daniloasfigueiredo** [comment]
Same problem here

**Comment by paulorpereirasouza** [comment]
Same

**Comment by hiran-filho** [comment]
The same here

**Comment by Gustavo-Marques19** [comment]
The same here

**Comment by evision-p** [comment]
Same Here

**Comment by nicolasmohamed** [comment]
Same here

`2023-01-20T06:56:34.9567360Z ##[error]Error: Error: Unrecognized argument 'Files\IIS\Microsoft'. All arguments must begin with "-".`

**Comment by penihel** [comment]
Same Here, too 👍

**Comment by penihel** [comment]
Same, but with this task (**ISWebAppDeploymentOnMachineGroup**)

See
```
023-01-20T10:13:08.5909595Z ##[section]Starting: Deploy Landing
2023-01-20T10:13:08.5999780Z ==============================================================================
2023-01-20T10:13:08.5999968Z Task : IIS web app deploy
2023-01-20T10:13:08.6000073Z Description : Deploy a website or web application using Web Deploy
2023-01-20T10:13:08.6000217Z Version : 0.216.0
2023-01-20T10:13:08.6000311Z Author : Microsoft Corporation
2023-01-20T10:13:08.6000419Z Help : https://docs.microsoft.com/azure/devops/pipelines/tasks/deploy/iis-web-app-deployment-on-machine-group
2023-01-20T10:13:08.6000683Z ==============================================================================

...
...
...



2023-01-20T10:13:09.4155735Z [command]“C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe” -verb:sync -source:package=‘C:\azagent\A1_work\r1\a\temp_web_package_4560503770922759.zip’ -dest:contentPath=‘my-site.net.br’ -skip:Directory=App_Data
2023-01-20T10:13:09.4811053Z ##[error]Failed to deploy web package to IIS website.
2023-01-20T10:13:09.4818865Z ##[error]Error: Unrecognized argument ‘Files\IIS\Microsoft’. All arguments must begin with “-”.
Error count: 1.

2023-01-20T10:13:09.4967700Z ##[error]Error: The process ‘C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe’ failed with exit code 4294967295
```

**Comment by leonardosimoura** [comment]
Same error with ISWebAppDeploymentOnMachineGroup here

**Comment by vmapetr** [comment]
@noellelc Thank you for reporting this issue. We are investigating it now.

**Comment by rfulgencio3** [comment]
Same problem

**Comment by felipefialho** [comment]
Is there any provision for normalization? It is impacting all deploys

**Comment by rodrigofudissaku** [comment]
Is there any workaround to this issue?

**Comment by tassioFront** [comment]
`workaround`: change App Deploy App Service task version. I've changed it to version 3 instead of 4.

**Comment by ericklodion** [comment]
Same here

**Comment by jopnine** [comment]
> `workaround`: change `App Deploy App Service` task version. I've changed it to version 3 instead of 4.

Where do i change `App Deploy App Service` version ?  I mean.. the path.

**Comment by ericklodion** [comment]
> > `workaround`: change `App Deploy App Service` task version. I've changed it to version 3 instead of 4.
> 
> Where do i change `App Deploy App Service` version ? I mean.. the path.

- Edit pipeline
- Click on "Azure app service"
- Change the first combo option from 4.* to 3.*

**Comment by rfulgencio3** [comment]
> > `workaround`: change `App Deploy App Service` task version. I've changed it to version 3 instead of 4.
> 
> Where do i change `App Deploy App Service` version ? I mean.. the path.

I think this can help you: Go to Releases > select the project > Edit > under the envinronment click in 'jobs, taks' > click in Deploy Azure App Service > in 'Task version' select 3.*

NOTE: I try this but this not made effect for my deploy

**Comment by ericklodion** [comment]
> > > `workaround`: change `App Deploy App Service` task version. I've changed it to version 3 instead of 4.
> > 
> > 
> > Where do i change `App Deploy App Service` version ? I mean.. the path.
> 
> I think this can help you: Go to Releases > select the project > Edit > under the envinronment click in 'jobs, taks' > click in Deploy Azure App Service > in 'Task version' select 3.*
> 
> NOTE: I try this but this not made effect for my deploy

Really ? for me works fine ...

**Comment by jopnine** [comment]
I see, in my case i made small confusion. Im using "IIS web app deploy" instead of "Azure App Service"

**Comment by rfulgencio3** [comment]
> 

At first attempt e just ran only the release process and in the console it still the 4.216.2 version, after this, a just ran again all pipeline process and at this time and now, the release works fine for this environment and 3.* version.

**Comment by rfulgencio3** [comment]
> > 
> 
> At first attempt e just ran only the release process and in the console it still the 4.216.2 version, after this, a just ran again all pipeline process and at this time and now, the release works fine for this environment and 3.* version.

It possible downgrade for an specific version? I cannot edit the YAML
Et. can i use this version: 4.214.0 instead 3.*?

**Comment by nicolasmohamed** [comment]
downgrade from Azure App Service Deploy 4.* to 3.* seems to work

**Comment by Rafael-Cassiano** [comment]
Same problem, even returning to 3.* is not working 
 - task: AzureRmWebAppDeployment@3

**Comment by marcopaivaa** [comment]
> downgrade from Azure App Service Deploy 4.* to 3.* seems to work

is there any way to downgrade to a minor version (like 4.215.2) instead of changing from 4 to 3 with yaml?

**Comment by rafjsouza** [comment]
Any solutions for IISWebAppDeploymentOnMachineGroup task? Our pipeline is not working because of that error

**Comment by felipeagrela** [comment]
Same here in IISWebAppDeploymentOnMachineGroup...

**Comment by gabriel-alves-ivoryit** [comment]
Same here! Currently unable to release new versions of my applications.

I sse that past executions of my release pipeline were using version 0.215.0 o the "IIS web app deploy" task. The failing one mentions version 0.216.0. That is what i see in the execution logs
In the portal, though, the **only** task version option available is 0.*. It is obviously unchanged.

The error is the exact one as other people's around here: 

`##[error]Error: Unrecognized argument 'Files\IIS\Microsoft'. All arguments must begin with "-".`

**Comment by Rafael-Cassiano** [comment]
Working with version 4. After leaving BUILD Image: windows-2019 and DEPLOY Image: windows-2022

**Comment by danton721** [comment]
Quick solution was to rollback to version 0.215.0.

In a self-hosted agent I did the following steps:

Inside agent folder: `./_work/_tasks/IISWebAppDeploymentOnMachineGroup[...]`

1. Make sure that you have this folder: "0.215.0" (older versions might also work)
2. Delete content of folder "0.216.0" (I would recommend backing up)
3. Copy the contents of folder "0.215.0" inside folder "0.216.0"

This temporary solution did work for me with "IIS web app deploy" with a  self-hosted agent

**Comment by felipeagrela** [comment]
> Same here in IISWebAppDeploymentOnMachineGroup...

It comes work again...
Seems that the problem version is not available anymore... 

Problem Version: 0.216.0
Working Version: 0.215.0

**Comment by vmapetr** [comment]
We have reverted the affected tasks to previous versions in ADO, while we working on hotfixes:

- `AzureRmWebAppDeployment` - `4.216.2` -> `4.214.0`
- `IISWebAppDeploymentOnMachineGroup` -`0.216.0` -> `0.215.0`

Sorry for the inconvenience.

**Comment by JuliusFraga** [comment]
> We have reverted the affected tasks to previous versions in ADO, while we working on hotfixes:
> 
> * `AzureRmWebAppDeployment` - `4.216.2` -> `4.214.0`
> * `IISWebAppDeploymentOnMachineGroup` -`0.216.0` -> `0.215.0`
> 
> Sorry for the inconvenience.

Working fine now, thanks @vmapetr

**Comment by penihel** [comment]
How can i force to use 0.215.0 ?

My Azure Agent continue downloading the 0.216.0 version

**Comment by sfrsm** [comment]
> 

The same problem here.
I don't have the version 0.215.0 in my server.
How can I dowloand this version of the agent?

**Comment by felipeagrela** [comment]
> How can i force to use 0.215.0 ?
> 
> My Azure Agent continue downloading the 0.216.0 version

Are you running a new pipeline or re-running the same build?

**Comment by felipeagrela** [comment]
> > 
> 
> The same problem here. I don't have the version 0.215.0 in my server. How can I dowloand this version of the agent?

Are you running a new pipeline or re-running the same build?

**Comment by sfrsm** [comment]
> > > 
> > 
> > 
> > The same problem here. I don't have the version 0.215.0 in my server. How can I dowloand this version of the agent?
> 
> Are you running a new pipeline or re-running the same build?

A new pipeline.

**Comment by felipeagrela** [comment]
> > > > 
> > > 
> > > 
> > > The same problem here. I don't have the version 0.215.0 in my server. How can I dowloand this version of the agent?
> > 
> > 
> > Are you running a new pipeline or re-running the same build?
> 
> A new pipeline.

Try delete your task folder that has 0.216 version...

**Comment by felipeagrela** [comment]
> > > > > > 
> > > > > 
> > > > > 
> > > > > The same problem here. I don't have the version 0.215.0 in my server. How can I dowloand this version of the agent?
> > > > 
> > > > 
> > > > Are you running a new pipeline or re-running the same build?
> > > 
> > > 
> > > A new pipeline.
> > 
> > 
> > Try delete your task folder that has 0.216 version...
> 
> It still download the version 0.216.

Its on premise server agent or cloud server?

**Comment by jopnine** [comment]
Managed to make it work by unistalling ms deploy v3 from my Windows.

**Comment by jopnine** [comment]
> Managed to make it work by unistalling ms deploy v3 from my Windows.

![image](https://user-images.githubusercontent.com/15256203/214131317-70df02c8-872a-408e-a6fd-eddc19eb59ee.png)

After unistalling it, everything went fine.

**Comment by sfrsm** [comment]
Here I have installed the Microsoft Web Deploy 4.0 installed. I think the problem is in some library that handle the command line, because if you copy and run it manually it works, in powershell and the command terminal.

I solved here the problem adding a new Task in my pipeline to execute the command line script.
I added the line that was failling in this script and it works fine.

**Comment by DarkChocoholicDev** [comment]
> We have reverted the affected tasks to previous versions in ADO, while we working on hotfixes:
> 
> * `AzureRmWebAppDeployment` - `4.216.2` -> `4.214.0`
> * `IISWebAppDeploymentOnMachineGroup` -`0.216.0` -> `0.215.0`
> 
> Sorry for the inconvenience.

I'm seeing this issue even today. Was the above supposed to somehow prevent 0.216.0 from continuing to be pulled down by agents? I keep deleting the directory, but it keeps coming back. If I clone 0.215.0 and rename 0.216.0, it seems to get modified during my job run, resulting in the same failure. None of the given work-arounds I've tried seem to be working. I'm not prepared to uninstall Microsoft Web Deploy from my public-facing systems, since I don't have a good understanding of the likely side effects.

Any further work-around ideas for this issue?

**Comment by sfrsm** [comment]
> 

Are you running it on premise, did you tried to run it manually?

If it is on premise, try add a new task in your pipeline, Command Line Script, and insert the line with the webdeploy.exe in the script.

![image](https://user-images.githubusercontent.com/6587137/214391780-4d50981b-1c2b-4c41-bb5d-6616cc78ab51.png)

**Comment by DarkChocoholicDev** [comment]
It doesn't seem that we can do that because the filename keeps changing, perhaps because we're doing various forms of config-file transformation via the deployment task. Here are excerpts from the command lines produced by the last three runs.

`"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package='C:\azagent\A1\_work\r1\a\temp_web_package_3922506065380085.zip' ...
"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package='C:\azagent\A1\_work\r1\a\temp_web_package_7567551291335439.zip' ...
"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package='C:\azagent\A1\_work\r1\a\temp_web_package_1927501900671098.zip'...`

**Comment by mselley** [comment]
having the same issue here what a nightmare.

**Comment by sfrsm** [comment]
> It doesn't seem that we can do that because the filename keeps changing, perhaps because we're doing various forms of config-file transformation via the deployment task. Here are excerpts from the command lines produced by the last three runs.
> 
> `"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package='C:\azagent\A1\_work\r1\a\temp_web_package_3922506065380085.zip' ...`
> 
> `"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package='C:\azagent\A1\_work\r1\a\temp_web_package_7567551291335439.zip' ...`
> 
> `"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package='C:\azagent\A1\_work\r1\a\temp_web_package_1927501900671098.zip'...`

In my case, I changed in the build pipeline to generate a .zip file with fixed name like WebApp.zip. I forgot to mention that.
In the Release pipeline, in the IIS Web App Deploy  I flagged the Control Options to Continue on error.
This workaround works here.

**Comment by SteGriff** [comment]
I had this problem when building with the Microsoft hosted agent (default) and deploying onto on-premises IIS servers.

### My workaround

On the machine in the deployment group, I went to the `IISWebAppDeploymentOnMachineGroup` task directory, at:
`C:\azagent\A2\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba`

I backed up the content of the `0.216.0` directory onto the desktop.

I deleted the content of the `0.216.0` directory 

I copied the content of the `0.215.0` directory into  `0.216.0`

This prevents the build agent from trying to re-download 0.216, and when it uses 0.216, it gets the 0.215 implementation instead, because that's what is in the folder now.

Hope it helps.

**Comment by arezn** [comment]
> I had this problem when building with the Microsoft hosted agent (default) and deploying onto on-premises IIS servers.
> 
> ### My workaround
> On the machine in the deployment group, I went to the `IISWebAppDeploymentOnMachineGroup` task directory, at: `C:\azagent\A2\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba`
> 
> I backed up the content of the `0.216.0` directory onto the desktop.
> 
> I deleted the content of the `0.216.0` directory
> 
> I copied the content of the `0.215.0` directory into `0.216.0`
> 
> This prevents the build agent from trying to re-download 0.216, and when it uses 0.216, it gets the 0.215 implementation instead, because that's what is in the folder now.
> 
> Hope it helps.

It works

**Comment by DarkChocoholicDev** [comment]
Ok, yes, indeed, it works! There was a subtlety, though, that I didn't realize, which made all the difference. We couldn't just delete 0.216.0, clone 0.215.0, and rename the clone to 0.216.0. Rather, we truly had to delete the contents of the existing 0.216.0 and then copy in the contents of 0.215.0. It seemed like they'd be equivalent, but it appears to not be. Maybe there is some type of directory timestamp checking going on or something. Or maybe it had to do with us also deleting the 0.216.0.xxx files in the root of C:\azagent\A2\_work\_tasks\IISWebAppDeploymentOnMachineGroup_xxx. Not sure. But once I _truly_ followed the work-around directions, it worked.

Thank you very much!

**Comment by SteGriff** [comment]
> Or maybe it had to do with us also deleting the 0.216.0.xxx files in the root of C:\azagent\A2_work_tasks\IISWebAppDeploymentOnMachineGroup_xxx. Not sure. But once I _truly_ followed the work-around directions, it worked.

I think you would have gotten away with doing it your way if you had not also deleted the `.completed` and `.node6` files, which appears to be a signal to DevOps agent that the version installed correctly at a given datetime. I'm guessing that if they're absent, it reinstalls the version.

Glad it worked! 😀

**Comment by AlexisMateo** [comment]
> I had this problem when building with the Microsoft hosted agent (default) and deploying onto on-premises IIS servers.
> 
> ### My workaround
> On the machine in the deployment group, I went to the `IISWebAppDeploymentOnMachineGroup` task directory, at: `C:\azagent\A2\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba`
> 
> I backed up the content of the `0.216.0` directory onto the desktop.
> 
> I deleted the content of the `0.216.0` directory
> 
> I copied the content of the `0.215.0` directory into `0.216.0`
> 
> This prevents the build agent from trying to re-download 0.216, and when it uses 0.216, it gets the 0.215 implementation instead, because that's what is in the folder now.
> 
> Hope it helps.

Its works for me

**Comment by jxsierra** [comment]
> I had this problem when building with the Microsoft hosted agent (default) and deploying onto on-premises IIS servers.
> 
> ### My workaround
> On the machine in the deployment group, I went to the `IISWebAppDeploymentOnMachineGroup` task directory, at: `C:\azagent\A2\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba`
> 
> I backed up the content of the `0.216.0` directory onto the desktop.
> 
> I deleted the content of the `0.216.0` directory
> 
> I copied the content of the `0.215.0` directory into `0.216.0`
> 
> This prevents the build agent from trying to re-download 0.216, and when it uses 0.216, it gets the 0.215 implementation instead, because that's what is in the folder now.
> 
> Hope it helps.

Thank you very much! Its works for me

**Comment by zulfiqarali310** [comment]
> We have reverted the affected tasks to previous versions in ADO, while we working on hotfixes:
> 
> * `AzureRmWebAppDeployment` - `4.216.2` -> `4.214.0`
> * `IISWebAppDeploymentOnMachineGroup` -`0.216.0` -> `0.215.0`
> 
> Sorry for the inconvenience.


Hello vmapetr, 
still facing the issue neither of the above solutions works for me, it's still pointing to 0.216.0 instead of 0.215.0, 
please let me know if got any solutions to this fix. Thanks


<img width="1225" alt="123" src="https://user-images.githubusercontent.com/62157842/214889276-36f8808d-d03d-4949-b445-249057b90c2d.png">

**Comment by DarkChocoholicDev** [comment]
zulfiqarali310, have you done the following steps _exactly_? It only worked for me when I did.

1. After a failure, bring up the folder ...\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba.
2. Without deleting the 0.216.0 folder, delete all contents of the 0.216.0 folder (files and subfolders).
3. Copy all of the contents of the 0.215.0 folder into the 0.216.0 folder.
4. Leave the files "0.216.0.completed" and "0.216.0.node6" as-is.
5. Retry your deployment stage (or create a new release).

**Comment by zulfiqarali310** [comment]
> zulfiqarali310, have you done the following steps _exactly_? It only worked for me when I did.
> 
> 1. After a failure, bring up the folder `...\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba`.
> 2. Without deleting the `0.216.0` folder itself, delete all contents of the `0.216.0` folder (files and subfolders).
> 3. Copy all of the contents of the `0.215.0` folder into the `0.216.0` folder.
> 4. Leave the files `0.216.0.completed` and `0.216.0.node6` as-is.
> 5. Retry your deployment stage (or create a new release).


Keith Frechette Thanks for your reply, 

Actually, i did create new release pipeline its getting failed, the old one is also falling,  non of the above solutions work for me. 

Actually, i am using cloud base server and deploying applications using Deploy IIS Website/App

Does anybody have solutions to it.? Thanks

**Comment by johnkamau** [comment]
Same issue as the other users above. Switching contents of the directory 0.216 with those of 0.215 on the target server resolved the existing release, not sure if a new release will work if I re-run it.

**Comment by zulfiqarali310** [comment]
> Same issue as the other users above. Switching contents of the directory 0.216 with those of 0.215 on the target server resolved the existing release, not sure if a new release will work if I re-run it.

not working for the new release, and i need it for a new release.

**Comment by johnkamau** [comment]
@zulfiqarali310 I tihnk they mean that you copy the files after the release has failed. Then re-deploy the release, the published version will be your latest release.

**Comment by zulfiqarali310** [comment]
> 

Thanks, everyone somehow it did work by doing this way.. thanks for your comments.

**Comment by zulfiqarali310** [comment]
> zulfiqarali310, have you done the following steps _exactly_? It only worked for me when I did.
> 
> 1. After a failure, bring up the folder `...\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba`.
> 2. Without deleting the `0.216.0` folder itself, delete all contents of the `0.216.0` folder (files and subfolders).
> 3. Copy all of the contents of the `0.215.0` folder into the `0.216.0` folder.
> 4. Leave the files `0.216.0.completed` and `0.216.0.node6` as-is.
> 5. Retry your deployment stage (or create a new release).


Thanks, everyone somehow it did work by doing this way.. thanks for your comments.

**Comment by amantcs** [comment]
Uninstalled the web deploy from the web server as we are doing deployment through pipelines only, resolved our issue.

**Comment by Kablekarnage** [comment]
Same issue.  We were able to fix this by uninstalling web deploy from the machine and clearing out the locally cached 'tasks' folder in the agent directory, e.g. 'C:\azagent\A4\_work\_tasks'

**Comment by enginhorzum** [comment]
I have the same problem. 
`Unrecognized argument 'Pipeline'. All arguments must begin with "-".`
It started after IISWebAppDeploymentOnMachineGroup task updated to version 0.216.0
But this error only occurs on our on-premise servers with a Pipeline Agent version 2.213.2 installed.
The servers, which have Pipeline Agent version 2.193.0 installed, work without any error with IISWebAppDeploymentOnMachineGroup task version 0.216.0

**Comment by esuyg** [comment]
Had the same issue. Uninstalling Web Deploy did fix the IISWebAppDeploymentOnMachineGroup task, but we need Web Deploy installed for the earlier MSBuild task so had to use @SteGriff's workaround for now - thanks!

**Comment by pablosguajardo** [comment]
the same problem in version: 0.216.0

**Comment by pablosguajardo** [comment]
> zulfiqarali310, have you done the following steps _exactly_? It only worked for me when I did.
> 
> 1. After a failure, bring up the folder `...\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba`.
> 2. Without deleting the `0.216.0` folder itself, delete all contents of the `0.216.0` folder (files and subfolders).
> 3. Copy all of the contents of the `0.215.0` folder into the `0.216.0` folder.
> 4. Leave the files `0.216.0.completed` and `0.216.0.node6` as-is.
> 5. Retry your deployment stage (or create a new release).

thanks with this work around was able to make it work!!!

**Comment by davidyardy** [comment]
This fix worked for me :)

**Comment by soc-martin** [comment]
The fix is related to commit 3193cff2fd6c5754cadb8b4dc1b0e0cc96f46350.

**Comment by aneudy91** [comment]
> 

Thank you so much, this help me fix this issue

**Comment by etaxi341** [comment]
Thank you, uninstalling Web Deploy and deleting the "_work" folders fixed it for me. The issue just started happening today. Yesterday everything was still working.

**Comment by Faiyaz-Aboobaker** [comment]
Thank you, uninstalling Web Deploy (V4) resolved the issue

**Comment by alexkads** [comment]
> Thank you, uninstalling Web Deploy (V4) resolved the issue

I just uninstalled it and WebDeploy worked. Thank you very much, because yesterday I didn't even sleep.

**Comment by paulahlers** [comment]
Was working yesterday for us, failed today.. removed webdeploy and it worked

**Comment by travis-morita** [comment]
Same here, uninstalling Web Deploy did the trick, without the need to delete any folders

**Comment by Kev8144** [comment]
The above temp solution of removing 0.216 contents worked for me except a couple things: 
1 - I didn't have version 0.215 to replace 0.216 contents so I used version 0.208
2 - the path for me was C:\azagent\A3\_work\_tasks\IISWebAppDeploymentOnMachineGroup

**Comment by SAPHemlock** [comment]
> I had this problem when building with the Microsoft hosted agent (default) and deploying onto on-premises IIS servers.
> 
> ### My workaround
> On the machine in the deployment group, I went to the `IISWebAppDeploymentOnMachineGroup` task directory, at: `C:\azagent\A2\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba`
> 
> I backed up the content of the `0.216.0` directory onto the desktop.
> 
> I deleted the content of the `0.216.0` directory
> 
> I copied the content of the `0.215.0` directory into `0.216.0`
> 
> This prevents the build agent from trying to re-download 0.216, and when it uses 0.216, it gets the 0.215 implementation instead, because that's what is in the folder now.
> 
> Hope it helps.

Thank you so much.
This workaround really fixed the problem.

**Comment by ozdreamer** [comment]
> I had this problem when building with the Microsoft hosted agent (default) and deploying onto on-premises IIS servers.
> 
> ### My workaround
> On the machine in the deployment group, I went to the `IISWebAppDeploymentOnMachineGroup` task directory, at: `C:\azagent\A2\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba`
> 
> I backed up the content of the `0.216.0` directory onto the desktop.
> 
> I deleted the content of the `0.216.0` directory
> 
> I copied the content of the `0.215.0` directory into `0.216.0`
> 
> This prevents the build agent from trying to re-download 0.216, and when it uses 0.216, it gets the 0.215 implementation instead, because that's what is in the folder now.
> 
> Hope it helps.

You are true champ bro. This trick saved my life in the middle of a prod deployment.

**Comment by MarkusPopp** [comment]
The workaround is not working in our case, because the tasks get downloaded fresh every deployment. Up to now, no way found to prevent the download, even if i set the version to 216 within the *.json files.

**Comment by patmoe** [comment]
The workaround worked.. :-) Thanks...

It looks like a classic bug..   read parameters and split on hyphen ( - )   
...but forgetting to skip content inside string deliiters.. 
My sitename had a hyphen in it - other logs shown in this issue also have paramvalues with hyphen inside..
jm2c

**Comment by patrice-sc** [comment]
Same here (likely when  deploying to folders with a - character). I saw the workaround but it really needs to be fixed please...

**Comment by sundar85** [comment]
same here, the workaround worked.

**Comment by MattJeanes** [comment]
The workaround has also worked for us too, although I am wondering why this is still an issue if it has supposedly rolled back? Is it due to us already having the latest version of the task installed already?

**Comment by daleua** [comment]
Same here ...someone forgot to put those quotes around the msdeploy command :(

**Comment by ruzena-kavanova** [comment]
Unistalling Web deploy fixed the issue for me. Thanks for help

**Comment by c0shea** [comment]
@Pranshu-Negi PR #17247 has broken IIS deployments. Any update on when this will be fixed?

**Comment by ngiessel** [comment]
Same here. Uninstalling Web Deploy from our server fixed it.

**Comment by SKO85** [comment]
We are using the `IISWebAppDeploymentOnMachineGroup` task that deploys the application to an Windows machine to an IIS setup. I can confirm that uninstalling `Web Deploy v4` from the server via `Apps & Features` worked in our case. Tested this on our testing machine. I assume the agent got updated and that this does not work well if Web Deploy was previously installed on the machine.

**Comment by kapilsharma749** [comment]
Uninstalling Web Deploy worked for me .

It is not Deploying the files at the location I needed with correct files from the package created(WebApp.zip)

**Comment by SimD80s** [comment]
A thing I noticed is that if you uninstall web deploy, it appears the task then runs the executable from 0.216.0\node_modules\azure-pipelines-tasks-webdeployment-common instead of from the local folder where Web deploy is installed in Program Files, and the  issue only arise in the latter case.

**Comment by caco0516** [comment]
> Uninstalled the web deploy 3.6 from the web server as we are doing deployment through pipelines only, resolved our issue. I am using Cloud-based Azure DevOps services and deploying into VM which hosts IIS, using IIS Web App Deploy task.

Thanks !!! this worked for me . Im using web deploy in a pipeline release stage.

**Comment by fernandofig** [comment]
> The workaround has also worked for us too, although I am wondering why this is still an issue if it has supposedly rolled back? Is it due to us already having the latest version of the task installed already?

This is a good point.

I tried deleting all the folders and files within _work\_tasks\IISWebAppDeploymentOnMachineGroup_*, restarted the Azure Agent service, and ran the stage again. As I expected, the task is redownloaded, but task that's download is still 0.216. So what gives? Wasn't it supposed to be 0.215?

**Comment by khuramshahzad** [comment]
Thank your for quick fix, i have uninstall the we deploy from machine (Web Deploy) and my CI-CD pipeline is working fine

**Comment by frjimenezcorpcab** [comment]
in my case, im new with azure pipelines and release, in my first try, i found this error, but when uninstall web deploy, release ocurrs without any error, but not deploy in the machine,  after try the command directly in power shell, seems like the error is the same that with web deploy  "C:\azagent\A5\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba\0.216.0\node_modules\azure-pipelines-tasks-webdeployment-common\MSDeploy3.6\MSDeploy3.6\msdeploy.exe" -verb:sync -source:package='C:\azagent\A5\_work\r1\a\temp_web_package_6798924466245622.zip' -dest:auto -setParam:name='IIS Web Application Name',value='RH' -enableRule:DoNotDeleteRule

**Comment by Pranshu-Negi** [comment]
The PR that created the issue has been reverted now.

**Comment by selangley-wa** [comment]
What they need to do is push out a newer version, e.g., 0.217.0, that works correctly.
It could just be 0.215.0 under a new name.

**Comment by edgariscoding** [comment]
I agree. As it is, clients with 0.216.0 will stay on that version unless a newer version is released.

**Comment by Pranshu-Negi** [comment]
The PR has been reverted but the fix will take few days to be deployed to production. Please be reassured that we are working to fix this issue as soon as possible.

**Comment by DarkChocoholicDev** [comment]
What happened, though? I thought that reversion was done several weeks ago.

**Comment by frjimenezcorpcab** [comment]
its any way to download 0.215.0  while release to production

**Comment by DarkChocoholicDev** [comment]
If nothing gets tailored to the installation when it's downloaded, in theory someone could ZIP it up and make it available on a share. I'm scoping it out now. 14.7 MB of files. (I guess efficiency does much come into play here. :-))
<img width="185" alt="image" src="https://user-images.githubusercontent.com/23127728/216685493-8b38dd12-5f1a-45da-8ddd-c24d6d838085.png">

**Comment by VIJAYASHANKAR** [comment]
We too facing this issue, and it's blocking the automated deployment process.

**Comment by MarkusPopp** [comment]
today i saw Task version 0.217.0 in the azagent work folder, which is actually 0.215.0, so the fix seems to be deployed into prod.

**Comment by jrummell** [comment]
As of an hour ago, this is now working for me

**Comment by engern** [comment]
Can confirm this also works for me, even with web deploy 4.0 installed. It downloaded a folder with version 0.217.0 before starting.

Still not impressed with how long it took to reverse a bug that potentially stops CI/CD for thousands of applications...

**Comment by EjazAtIces** [comment]
I am getting same error when I use Azure Devops Release pipeline to deploy to on-premises web server . This is IIS(asp.net core mvc) application.

Just to add more info: 
Deployment works if environment variable is set (ASPNETCORE_ENVIRONMENT)
If environment variable is deleted on the server then getting this error.

any help? any other page reference?
